Covid19Radar/Covid19Radar/Repository/DiagnosisKeyRepository.cs

```diff
@@ -0,0 +1,102 @@
+
+        public Task<string> DownloadDiagnosisKeysAsync(DiagnosisKeyEntry diagnosisKeyEntry, string outputDir, CancellationToken cancellationToken);
+    }
+
+    [JsonObject]
+    public class DiagnosisKeyEntry
+    {
+        [JsonProperty("region")]
+        public int Region;
+
+        [JsonProperty("url")]
+        public string Url;
+
+        [JsonProperty("created")]
+        public long Created;
+    }
+
+    public class DiagnosisKeyRepository : IDiagnosisKeyRepository
+    {
+        private const long BUFFER_LENGTH = 4 * 1024 * 1024;
+
+        private readonly HttpClient _client;
+        private readonly ILoggerService _loggerService;
+
+        public DiagnosisKeyRepository(
+            IHttpClientService httpClientService,
+            ILoggerService loggerService
+            )
+        {
+            _client = httpClientService.Create();
+            _loggerService = loggerService;
+        }
+
+        public async Task<IList<DiagnosisKeyEntry>> GetDiagnosisKeysListAsync(string url, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response = await _client.GetAsync(url, cancellationToken);
+            if (response.IsSuccessStatusCode)
+            {
+                string content = await response.Content.ReadAsStringAsync();
+                _loggerService.Debug(content);
+                return JsonConvert.DeserializeObject<List<DiagnosisKeyEntry>>(content);
+            }
+            else
+            {
+                _loggerService.Debug($"GetDiagnosisKeysListAsync {response.StatusCode}");
+            }
+
+            return new List<DiagnosisKeyEntry>();
+        }
+
+        public async Task<string> DownloadDiagnosisKeysAsync(DiagnosisKeyEntry diagnosisKeyEntry, string outputDir, CancellationToken cancellationToken)
+        {
+            Uri uri = new Uri(diagnosisKeyEntry.Url);
+            HttpResponseMessage response = await _client.GetAsync(uri, cancellationToken);
+            if (response.IsSuccessStatusCode)
+            {
+                string fileName = uri.Segments[uri.Segments.Length - 1];
+                string outputPath = Path.Combine(outputDir, fileName);
+
+                byte[] buffer = new byte[BUFFER_LENGTH];
+
+                using BufferedStream bs = new BufferedStream(await response.Content.ReadAsStreamAsync());
+                using FileStream fs = File.OpenWrite(outputPath);
+
+                int len = 0;
+                while ((len = await bs.ReadAsync(buffer, 0, buffer.Length)) > 0)
+                {
+                    await fs.WriteAsync(buffer, 0, len);
+                }
+
+                return outputPath;
+            }
+            else
+            {
+                _loggerService.Debug($"DownloadDiagnosisKeysAsync {response.StatusCode}");
+                throw new HttpRequestException($"Request is unsuccessful with status code {response.StatusCode}");
```

Comment:
[修正任意]
`unsuccessful` より `failed` のほうが一般的だと思いました。

```suggestion
                throw new HttpRequestException($"Request is failed with status code {response.StatusCode}");
```